// @antv/g2plot/lib/plots/dual-axes/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DualAxes = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var plot_1 = require("../../core/plot");
var utils_1 = require("../../utils");
var adaptor_1 = require("./adaptor");
var DualAxes = /** @class */ (function (_super) {
    tslib_1.__extends(DualAxes, _super);
    function DualAxes() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型: 双轴图 */
        _this.type = 'dual-axes';
        return _this;
    }
    /**
     * 获取 双轴图 默认配置
     */
    DualAxes.prototype.getDefaultOptions = function () {
        return utils_1.deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            yAxis: [],
            syncViewPadding: true,
        });
    };
    /**
     * 获取双轴图的适配器
     */
    DualAxes.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    /**
     * 设置状态
     * @param type 状态类型，支持 'active' | 'inactive' | 'selected' 三种
     * @param conditions 条件，支持数组
     * @param status 是否激活，默认 true
     */
    DualAxes.prototype.setState = function (type, condition, status) {
        if (status === void 0) { status = true; }
        var elements = utils_1.getAllElementsRecursively(this.chart);
        util_1.each(elements, function (ele) {
            if (condition(ele.getData())) {
                ele.setState(type, status);
            }
        });
    };
    /**
     * 获取状态
     */
    DualAxes.prototype.getStates = function () {
        var elements = utils_1.getAllElementsRecursively(this.chart);
        var stateObjects = [];
        util_1.each(elements, function (element) {
            var data = element.getData();
            var states = element.getStates();
            util_1.each(states, function (state) {
                stateObjects.push({ data: data, state: state, geometry: element.geometry, element: element });
            });
        });
        return stateObjects;
    };
    return DualAxes;
}(plot_1.Plot));
exports.DualAxes = DualAxes;
